mapper/thin_edge_json/src/json.rs

```diff
@@ -1,165 +1,250 @@
+}
+
+impl ThinEdgeJsonBuilder {
+    fn new() -> Self {
+        Self {
+            timestamp: None,
+            inside_group: None,
+            measurements: Vec::new(),
+        }
     }
 
-    pub fn from_str(
-        json_string: &str,
-        timestamp: DateTime<FixedOffset>,
-    ) -> Result<ThinEdgeJson, ThinEdgeJsonError> {
-        match json::parse(&json_string) {
-            Ok(thin_edge_obj) => ThinEdgeJson::from_json(thin_edge_obj, timestamp),
-            Err(err) => Err(ThinEdgeJsonError::new_invalid_json(json_string, err)),
+    fn finish(self) -> Result<ThinEdgeJson, ThinEdgeJsonError> {
+        assert!(self.inside_group.is_none());
+
+        if self.measurements.is_empty() {
+            return Err(ThinEdgeJsonError::EmptyThinEdgeJsonRoot);
+        }
+
+        Ok(ThinEdgeJson {
+            timestamp: self.timestamp,
+            values: self.measurements,
+        })
+    }
+}
+
+impl GroupedMeasurementVisitor for ThinEdgeJsonBuilder {
+    type Error = ThinEdgeJsonError;
+
+    fn timestamp(&mut self, value: DateTime<FixedOffset>) -> Result<(), Self::Error> {
+        match self.timestamp {
+            None => {
+                self.timestamp = Some(value);
+                Ok(())
+            }
+            Some(_) => Err(ThinEdgeJsonError::DuplicatedTimestamp),
         }
     }
 
-    ///Confirms that the json is in thin-edge json format or not
-    fn from_json(
-        input: json::JsonValue,
-        timestamp: DateTime<FixedOffset>,
-    ) -> Result<ThinEdgeJson, ThinEdgeJsonError> {
-        let mut measurements = vec![];
-        let mut time = timestamp;
-        match &input {
+    fn measurement(&mut self, name: &str, value: f64) -> Result<(), Self::Error> {
+        let measurement = SingleValueMeasurement::new(name, value)?;
+        if let Some(group) = &mut self.inside_group {
+            group.values.push(measurement);
+        } else {
+            self.measurements.push(ThinEdgeValue::Single(measurement));
+        }
+        Ok(())
+    }
+
+    fn start_group(&mut self, group: &str) -> Result<(), Self::Error> {
+        if self.inside_group.is_none() {
+            self.inside_group = Some(MultiValueMeasurement {
+                name: group.into(),
+                values: Vec::new(),
+            });
+            Ok(())
+        } else {
+            Err(ThinEdgeJsonError::UnexpectedStartOfGroup)
+        }
+    }
+
+    fn end_group(&mut self) -> Result<(), Self::Error> {
+        match self.inside_group.take() {
+            Some(group) => {
+                if group.values.is_empty() {
+                    return Err(ThinEdgeJsonError::EmptyThinEdgeJson { name: group.name });
+                } else {
+                    self.measurements.push(ThinEdgeValue::Multi(group))
+                }
+            }
+            None => return Err(ThinEdgeJsonError::UnexpectedEndOfGroup),
+        }
+        Ok(())
+    }
+}
+
+pub struct ThinEdgeJsonParser;

```

Comment:
This struct seems useless. Why not just a function `parse_thin_edge_json(input: &[u8], visitor: &mut T)` ?